sql/core/src/test/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListenerSuite.scala

```diff
@@ -852,6 +857,33 @@ class SQLAppStatusListenerSuite extends SharedSparkSession with JsonTestUtils
     assert(metrics.contains(expectedMetric.id))
     assert(metrics(expectedMetric.id) === expectedValue)
   }
+
+  test("SPARK-36030: Report metrics from Datasource v2 write") {
+    withTempDir { dir =>
+      val statusStore = spark.sharedState.statusStore
+      val oldCount = statusStore.executionsList().size
+
+      val cls = classOf[CustomMetricsDataSource].getName
+      spark.range(10).select('id as 'i, -'id as 'j).write.format(cls)
+        .option("path", dir.getCanonicalPath).mode("append").save()
+
+      // Wait until the new execution is started and being tracked.
+      while (statusStore.executionsCount() < oldCount) {
+        Thread.sleep(100)
+      }
+
+      // Wait for listener to finish computing the metrics for the execution.
+      while (statusStore.executionsList().isEmpty ||
+        statusStore.executionsList().last.metricValues == null) {
```

Comment:
ditto. This could be an infinite loop when the bug happens. Shall we use `eventually` style instead?